pkg/plugins/policies/meshtcproute/plugin/v1alpha1/listeners.go

```diff
@@ -0,0 +1,100 @@
+package v1alpha1
+
+import (
+	"github.com/pkg/errors"
+
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	core_xds "github.com/kumahq/kuma/pkg/core/xds"
+	meshhttproute_api "github.com/kumahq/kuma/pkg/plugins/policies/meshhttproute/api/v1alpha1"
+	plugins_xds "github.com/kumahq/kuma/pkg/plugins/policies/xds"
+	meshroute_xds "github.com/kumahq/kuma/pkg/plugins/policies/xds/meshroute"
+	envoy_common "github.com/kumahq/kuma/pkg/xds/envoy"
+	envoy_listeners "github.com/kumahq/kuma/pkg/xds/envoy/listeners"
+	envoy_names "github.com/kumahq/kuma/pkg/xds/envoy/names"
+	envoy_tags "github.com/kumahq/kuma/pkg/xds/envoy/tags"
+	"github.com/kumahq/kuma/pkg/xds/generator"
+)
+
+func generateListeners(
+	proxy *core_xds.Proxy,
+	toRulesTCP core_xds.Rules,
+	servicesAccumulator envoy_common.ServicesAccumulator,
+) (*core_xds.ResourceSet, error) {
+	resources := core_xds.NewResourceSet()
+	// Cluster cache protects us from creating excessive amount of clusters.
+	// For one outbound we pick one traffic route, so LB and Timeout are
+	// the same.
+	clusterCache := map[string]struct{}{}
+	sc := &meshroute_xds.SplitCounter{}
+	networking := proxy.Dataplane.Spec.GetNetworking()
+	routing := proxy.Routing
+	toRulesHTTP := proxy.Policies.Dynamic[meshhttproute_api.MeshHTTPRouteType].
+		ToRules.Rules
+
+	for _, outbound := range networking.GetOutbound() {
+		tags := outbound.GetTagsIncludingLegacy()
+		serviceName := tags[mesh_proto.ServiceTag]
+		protocol := plugins_xds.InferProtocol(routing, serviceName)
+
+		backendRefs := getBackendRefs(toRulesTCP, toRulesHTTP, serviceName,
+			protocol)
+		if len(backendRefs) == 0 {
+			continue
+		}
```

Comment:
With ComputeConf we probably don't need to create separate `getBackendRefs` and `getTCPBackendRefs` functions:
```suggestion
		if matchingHTTPRuleExist(httpRules, serviceName, protocol) {
			continue
		}
		conf := core_xds.ComputeConf(toRulesTCP, core_xds.MeshService(serviceName))
		if conf == nil {
			continue
		}
		backendRefs := conf.Default.BackendRef // no need to check on nil, should be handled by validation
```

Reply:
I like to have it in separate function as it's returning some nested values because it's easier for me to read overall code. Please reopen the conversation if you strongly disagree.